### PR TITLE
Split Acme TEST implementation into separate hex component

### PR DESCRIPTION
**Pull request by aforward**
First, thanks for the great library.

As I was understanding the inner workings, I thought it would valuable to separate out the test ACME server into a separate library, but keep the integration with site_encrypt relatively untouched.

The server code has been pulled out into

https://hex.pm/packages/acme_ex

I have added tests to the implementation, but am not familiar (yet) with @spec annotations so I left those out.

In this PR, I have added the dependency on `:acme_ex` and then ripped out the related code.

Here's the output from running against the phoenix_demo project

```
[info] Running local ACME server at port 4002
[info] Running PhoenixDemoWeb.Endpoint with Cowboy using http://0.0.0.0:4000
[info] Running PhoenixDemoWeb.Endpoint with Cowboy using https://0.0.0.0:4001
[info] Starting certbot
[info] GET /directory
[info] HEAD /new
[info] POST /new-order
[info] GET /authorizations/2/3
[info] POST /challenge/http/2/3
[info] Sending challenge [Gml6Tagy3Mq1A0BJ8lzTVg] to localhost:4000
[info] Sending challenge [Gml6Tagy3Mq1A0BJ8lzTVg] to localhost:4000
[info] Sending challenge [Gml6Tagy3Mq1A0BJ8lzTVg] to localhost:4000
[info] GET /.well-known/acme-challenge/Gml6Tagy3Mq1A0BJ8lzTVg
[info] GET /.well-known/acme-challenge/Gml6Tagy3Mq1A0BJ8lzTVg
[info] GET /.well-known/acme-challenge/Gml6Tagy3Mq1A0BJ8lzTVg
[info] Sent 200 in 1ms
[info] Sent 200 in 1ms
[info] Sent 200 in 1ms
18:05:40 - info: compiled 6 files into 2 files, copied 3 in 1.1 sec
[info] GET /authorizations/2/3
[info] POST /finalize/2/3
[info] GET /cert/2/3
[info] Saving debug log to /Users/aforward/sin/projects/oss/phoenix_demo_new/_build/dev/lib/phoenix_demo/priv/certbot/log/letsencrypt.log

-------------------------------------------------------------------------------
Processing
/Users/aforward/sin/projects/oss/phoenix_demo_new/_build/dev/lib/phoenix_demo/priv/certbot/config/renewal/foo.bar.conf
-------------------------------------------------------------------------------
Cert is due for renewal, auto-renewing...
Plugins selected: Authenticator webroot, Installer None
Renewing an existing certificate
Performing the following challenges:
http-01 challenge for localhost
Waiting for verification...
Cleaning up challenges

-------------------------------------------------------------------------------
new certificate deployed without reload, fullchain is
/Users/aforward/sin/projects/oss/phoenix_demo_new/_build/dev/lib/phoenix_demo/priv/certbot/config/live/foo.bar/fullchain.pem
-------------------------------------------------------------------------------

-------------------------------------------------------------------------------

Congratulations, all renewals succeeded. The following certs have been renewed:
  /Users/aforward/sin/projects/oss/phoenix_demo_new/_build/dev/lib/phoenix_demo/priv/certbot/config/live/foo.bar/fullchain.pem (success)
-------------------------------------------------------------------------------

[info] Obtained new certificate, restarting endpoint
[info] Running PhoenixDemoWeb.Endpoint with Cowboy using http://0.0.0.0:4000
[info] Running PhoenixDemoWeb.Endpoint with Cowboy using https://0.0.0.0:4001
[info] Certbot finished
```

### CODE REVIEW

File: lib/site_encrypt/certifier.ex [hunk 1]
```diff
@@ -27,7 +27,7 @@ defmodule SiteEncrypt.Certifier do
   def handle_info(other, state), do: super(other, state)
 
   @impl Parent.GenServer
-  def handle_child_terminated(:fetcher, _pid, _reason, state) do
+  def handle_child_terminated(:fetcher, _meta, _pid, _reason, state) do
```
Comment (aforward) [inline lib/site_encrypt/certifier.ex]: Based on the following warning from the latest parent project

```
warning: got "@impl Parent.GenServer" for function handle_child_terminated/4 but this behaviour does not specify such callback. The known callbacks are:

  * GenServer.code_change/3 (function)
  * GenServer.format_status/2 (function)
  * GenServer.handle_call/3 (function)
  * GenServer.handle_cast/2 (function)
  * Parent.GenServer.handle_child_terminated/5 (function)
  * GenServer.handle_info/2 (function)
  * GenServer.init/1 (function)
  * GenServer.terminate/2 (function)

  lib/site_encrypt/certifier.ex:30
```

File: lib/site_encrypt/phoenix.ex [hunk 2]
```diff
@@ -28,21 +28,11 @@ defmodule SiteEncrypt.Phoenix do
...
-  defp acme_server_adapter_spec(Plug.Adapters.Cowboy, port),
-    do: {Plug.Adapters.Cowboy, scheme: :http, options: [port: port, acceptors: 1]}
-
-  defp dns(config, endpoint) do
-    [config.domain | config.extra_domains]
-    |> Enum.map(&{&1, fn -> "localhost:#{endpoint.config(:http) |> Keyword.fetch!(:port)}" end})
-    |> Enum.into(%{})
+    config
+    |> Map.put(:dns, AcmeEx.Endpoint.dns(endpoint, config))
+    |> AcmeEx.Standalone.child_spec()
```
Comment (aforward) [inline lib/site_encrypt/phoenix.ex]: Here's the main swap out; based on the library

https://github.com/aforward/acme_ex

**Comment by sasa1977** [comment]
Thank you for the pull request!

I have a lot of ideas for this library, and not a lot of time, so any help is appreciated. From what I can tell, a lot of work has been invested into `acme_ex`. For both of these reasons, I'm really sad that I have to reject this PR.

Unfortunately, I can't accept it for two important reasons:

1. I think it's too early to separate this project into two libraries.
2. `acme_ex` is a complete rewrite of the `AcmeServer` part of this library, and it changes semantics, guarantees, and some foundations I've laid with this project.

Here's a more elaborate explanation.

## Project separation

I agree that there might be potential to separate this project into two libraries, and in fact this is why I immediately placed `AcmeServer` in a separate namespace. However, I think it's way too early for such split. The reason is that I expect some possible changes in the interface of `AcmeServer`, and if this is done in a separate project, it will be much harder to iterate.

For example, one of the first things I want to tackle next is add tests for the client part of `SiteEncrypt`. This is one of the major reasons why I created `AcmeServer`. That will likely require some changes to the server interface, so I can instrument the server from the tests. If we do this with `acme_ex`, then I need to fork that project, make a PR, wait for it to be merged, and then wait for the hex package to be published. That's a huge amount of ceremony, which I don't want to introduce so early in the project.

## Rewrite of `AcmeServer`

For the reasons which I can't understand, the `acme_ex` project is a complete rewrite of `AcmeServer` part. Unfortunately, that makes it very hard for me to asses how much of the original foundations have been preserved. I basically have to reverse engineer your complete solution, then load all the details of the original into my brain, and then make a comparison. My brain capacity is not sufficient for this task.

I took a quick glance, and already I found two changes which I'm not willing to approve at the moment:

- coupling of the server flow to plug
- fault-tolerance of pending challenge requests

### Coupling of the server flow to plug

Currently, [the server flow is decoupled from plug](https://github.com/sasa1977/site_encrypt/blob/a78c0e9723fe7c1d3df0c1e3f3a64a41990662e7/lib/acme_server.ex). This is a deliberate decision which paves the foundation for supporting other servers, such as plain cowboy,  [raxx](https://github.com/CrowdHailer/raxx), or anything else that people will use in the Elixir and Erlang community. I also feel that such design supports testability, because we don't need to test through plug conn.

In contrast, `acme_ex` [couples the flow with the plug router](https://github.com/aforward/acme_ex/blob/45fe49a4839e34a2bb2776b1093cc7ff532ced6f/lib/router.ex), and by doing this, it essentially reverts that decisions without any prior discussion.

### Fault-tolerance of pending challenge requests

`AcmeServer` uses [a parent GenServer to manage a single challenge request](https://github.com/sasa1977/site_encrypt/blob/a78c0e9723fe7c1d3df0c1e3f3a64a41990662e7/lib/acme_server/jobs/http_verifier.ex). This is a deliberate decision which supports proper fault-tolerance. Since [each job instance is a temporary child](https://github.com/sasa1977/site_encrypt/blob/a78c0e9723fe7c1d3df0c1e3f3a64a41990662e7/lib/acme_server/jobs/http_verifier.ex#L2), it's guaranteed that a repeated failure of one challenge job will not take down other challenges.

In contrast, [acme_ex seems to use supervision restarts to handle failing requests](https://github.com/aforward/acme_ex/blob/45fe49a4839e34a2bb2776b1093cc7ff532ced6f/lib/challenge.ex#L6). 

First of all, I think that the current implementation here is wrong. You're setting the restart strategy for the task supervisor, not its children. Therefore, if a challenge job fails, it won't be restarted (because by the default `Task.Supervisor.start_child/3` starts the child with `restart: :temporary`). 

You probably want to set the restart attribute for each child, using [Task.Supervisor.start_child/3](https://hexdocs.pm/elixir/Task.Supervisor.html#start_child/3). Note, however, that in such implementation, three independent simultaneous failures will take down all currently running challenge jobs. 

More importantly, I believe that network failure is not a kind of thing which should be handled by supervisor restart. Supervisors are all about handling unexpected crashes, while network failure is definitely an expected situation. 

By using a parent GenServer, I layed the foundations to properly deal with networking failures, and separate failures of different challenges. Granted, the current implementation is unfinished and rough around the edges, but any deficiencies can be solved by extending it, not by rewriting it.

Again, `acme_ex` completely removes that decision without any prior discussion.

## Collaborating to SiteEncrypt

I've made this library because I needed it, but also because I think it serves as a nice demonstration of how simple development can be when working in BEAM. I think it can definitely help setting up Let's Encrypt certification. I have some ideas for extending and improving it, and it obviously needs tests and docs. Since my time is very limited, I can't really make it on my own, and therefore any help and contribution is welcome and much appreciated.

There are definitely some parts in `acme_ex` which I'd like to have in `AcmeServer`, most notably automated tests. However, I think such contributions should be made as focused PRs, not as a complete rewrite which, intentionally or not, reverts some decisions. 

If you feel that some of these decisions are not good, or you're maybe confused with the implementation, please open up an issue, so we can discuss it first. Perhaps we find out that some decisions are not optimal, or perhaps we find out that some decisions need to be explained in the comments, or maybe we discover that my implementation is overly complex and confusing. Either way, we can then move forward and address underlying issues properly.

## Going forward

Since I'm not willing to merge `acme_ex`, we're now left with two implementations. I deliberately use MIT license to grant as much freedom as possible to others. If you did use a significant portion of my code, I think you should have the `SiteEncrypt` license somewhere in `acme_ex`, but other than that you're free to do whatever you wish with the code.

I'm not really sure that we need multiple implementations of the server, but I definitely don't mind it, especially if you don't agree with some of the decisions I've made in `AcmeServer`. Ultimately, I'd like to have a well-tested plug-and-play solution for self-certification which requires no external processes or docker images. If you feel you can take it further with `acme_ex`, then definitely go with it. I'll be watching how it develops, and reconsider using it in this project in the future. Bot for the moment, I'll stick to the existing `AcmeServer` :-)

**Comment by aforward** [comment]
Understood.

I like to re-write code as a means to understand it, and to see if the architecture that emerges from a test-driven style aligns with the original.  The intent of the PR to get elicit feedback (as you have done) and to see the path forward.

From a conversation standpoint, I did not understand the use of the `Parent` supervisor and the role of the `Jobs` children for the server (the first area where our code diverges significantly). 

I will look to contribute smaller PRs to this project to help increase it's testing.

Finally, as for the license attribution; I am not sure what you mean by `I think you should have the SiteEncrypt license somewhere in acme_ex`.  I have a reference in the README to this project, but let me know where else you would like attribution.

**Pull request closed without merging**